README.md
# NAME

CtrlO::Crypt::XkcdPassword - Yet another xkcd style password generator

# VERSION

version 1.009

# SYNOPSIS

    use CtrlO::Crypt::XkcdPassword;
    my $password_generator = CtrlO::Crypt::XkcdPassword->new;

    say $password_generator->xkcd;
    # LimousineAllegeClergymanEconomic

    say $password_generator->xkcd( words => 3 );
    # ObservantFiresideMacho

    say $password_generator->xkcd( words => 3, digits => 3 );
    # PowerfulSpreadScarf645

    # Use custom word list
    CtrlO::Crypt::XkcdPassword->new(
      wordlist => '/path/to/wordlist'
    );
    CtrlO::Crypt::XkcdPassword->new(
      wordlist => 'Some::Wordlist::From::CPAN'
    );

    # Use another source of randomness (aka entropy)
    CtrlO::Crypt::XkcdPassword->new(
      entropy => Data::Entropy::Source->new( ... );
    );

# DESCRIPTION

`CtrlO::Crypt::XkcdPassword` generates a random password using the
algorithm suggested in [https://xkcd.com/936/](https://xkcd.com/936/): It selects 4 words
from a curated list of words and combines them into a hopefully easy
to remember password (actually a passphrase, but we're all trying to
get things done, so who cares..).

See [this
explaination](https://www.explainxkcd.com/wiki/index.php/936:_Password_Strength)
for detailed information on the security of passwords generated from a
known word list.

But [https://xkcd.com/927/](https://xkcd.com/927/) also applies to this module, as there are
already a lot of modules on CPAN implementing
[https://xkcd.com/936/](https://xkcd.com/936/). We still wrote a new one, mainly because we
wanted to use a strong source of entropy and a fine-tuned word list.

# METHODS

## new

    my $pw_generator = CtrlO::Crypt::XkcdPassword->new;

Initialize a new object. Uses `CtrlO::Crypt::XkcdPassword::Wordlist::en_gb`
as a word list per default. The default entropy is based on
`Crypt::URandom`, i.e. `/dev/urandom` and should be random enough (at
least more random than plain old `rand()`).

If you want / need to supply another source of entropy, you can do so
by setting up an instance of `Data::Entropy::Source` and passing it
to `new` as `entropy`.

    my $pw_generator = CtrlO::Crypt::XkcdPassword->new(
        entropy => Data::Entropy::Source->new( ... )
    );

To use one of the included language-specific word lists, do:

    my $pw_generator = CtrlO::Crypt::XkcdPassword->new(
        language => 'en-GB',
    );

Available languages are:

- en-GB

You can also provide your own custom word list, either in a file:

    my $pw_generator = CtrlO::Crypt::XkcdPassword->new(
        wordlist => '/path/to/file'
    );

Or in a module:

    my $pw_generator = CtrlO::Crypt::XkcdPassword->new(
        wordlist => 'My::Wordlist'
    );

See ["DEFINING CUSTOM WORD LISTS"](#defining-custom-word-lists) for more info.

## xkcd

    my $pw = $pw_generator->xkcd;
    my $pw = $pw_generator->xkcd( words  => 3 );
    my $pw = $pw_generator->xkcd( digits => 2 );

Generate a random, xkcd-style password.

Per default will return 4 randomly chosen words from the word list,
each word's first letter turned to upper case, and concatenated
together into one string:

    $pw_generator->xkcd;
    # CorrectHorseBatteryStaple

You can get a different number of words by passing in `words`. But
remember that anything smaller than 3 will probably make for rather
poor passwords, and anything bigger than 7 will be hard to remember.

You can also pass in `digits` to append a random number consisting of
`digits` digits to the password:

    $pw_generator->xkcd( words => 3, digits => 2 );
    # StapleBatteryCorrect75

# DEFINING CUSTOM WORD LISTS

Please note that `language` is only supported for the word lists
included in this distribution.

## in a plain file

Put your word list into a plain file, one line per word. Install this
file somewhere on your system. You can now use your word list like
this:

    CtrlO::Crypt::XkcdPassword->new(
      wordlist => '/path/to/wordlist'
    );

## in a Perl module using the Wordlist API

[Perlancar](https://metacpan.org/author/PERLANCAR) came up with a unified API for various word list modules,
implemented in [Wordlist](https://metacpan.org/pod/WordList). Pack
your list into a module adhering to this API, install the module, and
load your word list:

    CtrlO::Crypt::XkcdPassword->new(
      wordlist => 'Your::Cool::Wordlist'
    );

You can check out [CtrlO::Crypt::XkcdPassword::Wordlist::en\_gb](https://metacpan.org/pod/CtrlO%3A%3ACrypt%3A%3AXkcdPassword%3A%3AWordlist%3A%3Aen_gb) (included in
this distribution) for an example. But it's really quite simple: Just
subclass `Wordlist` and put your list of words into the `__DATA__`
section of the module, one line per word.

## in a Perl module using the Crypt::Diceware API

David Golden uses a different API in his [Crypt::Diceware](https://metacpan.org/pod/Crypt%3A%3ADiceware) module,
which inspired the design of [CtrlO::Crypt::XkcdPassword](https://metacpan.org/pod/CtrlO%3A%3ACrypt%3A%3AXkcdPassword). To use one
of those word lists, use:

    CtrlO::Crypt::XkcdPassword->new(
      wordlist => 'Crypt::Diceware::Wordlist::Common'
    );

(yes, this looks just like when using `Wordlist`. We inspect the
wordlist module and try to figure out what kind of API you're using)

To create a module using the [Crypt::Diceware](https://metacpan.org/pod/Crypt%3A%3ADiceware) wordlist API, just
create a package containing a public array `@Words` containing your
word list.

# WRAPPER SCRIPT

This distributions includes a simple wrapper script, [pwgen-xkcd.pl](https://metacpan.org/pod/pwgen-xkcd.pl).

# RUNNING FROM GIT

This is **not** the recommended way to install / use this module. But
it's handy if you want to submit a patch or play around with the code
prior to a proper installation.

## Carton

    git clone git@github.com:domm/CtrlO-Crypt-XkcdPassword.git
    carton install
    carton exec perl -Ilib -MCtrlO::Crypt::XkcdPassword -E 'say CtrlO::Crypt::XkcdPassword->new->xkcd'

## cpanm & local::lib

    git clone git@github.com:domm/CtrlO-Crypt-XkcdPassword.git
    cpanm -L local --installdeps .
    perl -Mlocal::lib=local -Ilib -MCtrlO::Crypt::XkcdPassword -E 'say CtrlO::Crypt::XkcdPassword->new->xkcd'

# SEE ALSO

Inspired by [https://xkcd.com/936/](https://xkcd.com/936/) and [https://xkcd.com/927/](https://xkcd.com/927/).

There are a lot of similar modules on CPAN, so we just point you to
[Neil Bower's comparison of CPAN modules for generating passwords](http://neilb.org/reviews/passwords.html).

## But why did we write yet another module?

- Good entropy

    Most of the password generating modules just use `rand()`, which "is
    not cryptographically secure" (according to perldoc).
    `CtrlO::Crypt::XkcdPassword` uses [Crypt::URandom](https://metacpan.org/pod/Crypt%3A%3AURandom) via
    [Data::Entropy](https://metacpan.org/pod/Data%3A%3AEntropy), which provides good entropy while still being portable.

- Good word list

    While [Crypt::Diceware](https://metacpan.org/pod/Crypt%3A%3ADiceware) has good entropy, we did not like its word
    lists. Of course we could have just provided a word list better suited
    to our needs, but we wanted it to be very easy to generate xkcd-style
    passwords.

- Easy API

    `my $pwd = CtrlO::Crypt::XkcdPassword->new->xkcd` returns 4 words
    starting with an uppercase letter as a string, which is our main use
    case. Nevertheless, the API also allows for more or fewer words, or
    even some digits.

- Fork save
- [https://xkcd.com/927/](https://xkcd.com/927/)

# THANKS

- Thanks to [Ctrl O](http://www.ctrlo.com/) for funding the development of this module.
- We learned the usage of `Data::Entropy` from
[https://metacpan.org/pod/Crypt::Diceware](https://metacpan.org/pod/Crypt::Diceware), which also implements an
algorithm to generate a random passphrase.
- [m\_ueberall](https://twitter.com/m_ueberall/status/965263922310909952)
for pointing out
[https://www.explainxkcd.com/wiki/index.php/936:\_Password\_Strength](https://www.explainxkcd.com/wiki/index.php/936:_Password_Strength)

# AUTHOR

Thomas Klausner <domm@plix.at>

# COPYRIGHT AND LICENSE

This software is copyright (c) 2018 - 2021 by Thomas Klausner.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.


Changes
Revision history for CtrlO-Crypt-XkcdPassword

{{$NEXT}}

1.011     2021-11-11 22:11:31+01:00
    - pass wordlist via commandline
    - add various EFF wordlists


1.010     2021-04-04 16:41:54+02:00
    - Add initial GitHub actions config (Paul Cochrane)
    - Set minimum Perl version explicitly (Paul Cochrane)
    - Fix typos, docs, remove whitespace (Paul Cochrane)
    - Add missing `dist.ini` attributes (Paul Cochrane)


1.009     2021-01-03 17:23:09+01:00
    - dist housekeeping: github issues,
      Dist::Zilla::PluginBundle::Author::DOMM


1.008   2020-04-30T14:34:00
    - Avoid generating negative integers (Tom Hukins)


1.007   2020-04-29T21:40:43
    - Fix pwgen-xkcd.pl when called with no arguments (Tom Hukins)


1.006	2019-08-07T19:19:20
    - Fixed pod error reported by CPANTS. (Mohammad S Anwar)


1.005   2018-04-02T21:36:04
    - improve error-tests
    - Fixed broken link in docs, reported by manwar via
      https://github.com/domm/CtrlO-Crypt-XkcdPassword/pull/2
    - Added params check for method xkcd() with unit tests.
      (Mohammad S Anwar)


1.004   2018-03-19T10:28:04
    - add author test in xt to check for bad words
    - remove a few more bad-ish words


1.003   2018-02-28T11:17:15
    - Remove words with potential to generate offensive phrases (Andy Beverley)


1.002   2018-02-19T19:21:27
    - fix copyright year (spotted by koki)
    - add link to https://www.explainxkcd.com/wiki/index.php/936:_Password_Strength (suggested by twitter.com/m_ueberall)
    - skip t/40-fork.t on MSWin32 (uncovered by CPAN testers)


1.001   2018-02-18T14:31:05
    - minor documentation improvements


1.000   2018-02-18T14:01:17
    - first release to CPAN



lib/CtrlO/Crypt/XkcdPassword/Wordlist/eff_short_1.pm
package CtrlO::Crypt::XkcdPassword::Wordlist::eff_short_1;
use strict;
use warnings;
use utf8;
use 5.010;

use base 'WordList';

# ABSTRACT: word list for CtrlO::Crypt::XkcdPassword based on eff_short_1_wordlist.txt

# VERSION

=head1 DESCRIPTION

Subclass of L<Wordlist> providing the a word list for
CtrlO::Crypt::XkcdPassword based on EFF's general short word list

Please see L<Wordlist> for the API if ever need to use this word list
in another context.

=head2 Wordlist source

=over

=item * L<https://www.eff.org/files/2016/07/18/eff_short_1_wordlist.txt>

=item * Please read L<EFF's New Wordlists for Random Passphrases|https://www.eff.org/deeplinks/2016/07/new-wordlists-random-passphrases> for the pros and cons of using this or other wordlists provided by EFF.

=back

=cut

'correct horse battery staple';

__DATA__
acid
acorn
acre
acts
afar
affix
aged
agent
agile
aging
agony
ahead
aide
aids
aim
ajar
alarm
alias
alibi
alien
alike
alive
aloe
aloft
aloha
alone
amend
amino
ample
amuse
angel
anger
angle
ankle
apple
april
apron
aqua
area
arena
argue
arise
armed
armor
army
aroma
array
arson
art
ashen
ashes
atlas
atom
attic
audio
avert
avoid
awake
award
awoke
axis
bacon
badge
bagel
baggy
baked
baker
balmy
banjo
barge
barn
bash
basil
bask
batch
bath
baton
bats
blade
blank
blast
blaze
bleak
blend
bless
blimp
blink
bloat
blob
blog
blot
blunt
blurt
blush
boast
boat
body
boil
bok
bolt
boned
boney
bonus
bony
book
booth
boots
boss
botch
both
boxer
breed
bribe
brick
bride
brim
bring
brink
brisk
broad
broil
broke
brook
broom
brush
buck
bud
buggy
bulge
bulk
bully
bunch
bunny
bunt
bush
bust
busy
buzz
cable
cache
cadet
cage
cake
calm
cameo
canal
candy
cane
canon
cape
card
cargo
carol
carry
carve
case
cash
cause
cedar
chain
chair
chant
chaos
charm
chase
cheek
cheer
chef
chess
chest
chew
chief
chili
chill
chip
chomp
chop
chow
chuck
chump
chunk
churn
chute
cider
cinch
city
civic
civil
clad
claim
clamp
clap
clash
clasp
class
claw
clay
clean
clear
cleat
cleft
clerk
click
cling
clink
clip
cloak
clock
clone
cloth
cloud
clump
coach
coast
coat
cod
coil
coke
cola
cold
colt
coma
come
comic
comma
cone
cope
copy
coral
cork
cost
cot
couch
cough
cover
cozy
craft
cramp
crane
crank
crate
crave
crawl
crazy
creme
crepe
crept
crib
cried
crisp
crook
crop
cross
crowd
crown
crumb
crush
crust
cub
cult
cupid
cure
curl
curry
curse
curve
curvy
cushy
cut
cycle
dab
dad
daily
dairy
daisy
dance
dandy
darn
dart
dash
data
date
dawn
deaf
deal
dean
debit
debt
debug
decaf
decal
decay
deck
decor
decoy
deed
delay
denim
dense
dent
depth
derby
desk
dial
diary
dice
dig
dill
dime
dimly
diner
dingy
disco
dish
disk
ditch
ditzy
dizzy
dock
dodge
doing
doll
dome
donor
donut
dose
dot
dove
down
dowry
doze
drab
drama
drank
draw
dress
dried
drift
drill
drive
drone
droop
drove
drown
drum
dry
duck
duct
dude
dug
duke
duo
dusk
dust
duty
dwarf
dwell
eagle
early
earth
easel
east
eaten
eats
ebay
ebony
ebook
echo
edge
eel
eject
elbow
elder
elf
elk
elm
elope
elude
elves
email
emit
empty
emu
enter
entry
envoy
equal
erase
error
erupt
essay
etch
evade
even
evict
evil
evoke
exact
exit
fable
faced
fact
fade
fall
false
fancy
fang
fax
feast
feed
femur
fence
fend
ferry
fetal
fetch
fever
fiber
fifth
fifty
film
filth
final
finch
fit
five
flag
flaky
flame
flap
flask
fled
flick
fling
flint
flip
flirt
float
flock
flop
floss
flyer
foam
foe
fog
foil
folic
folk
food
fool
found
fox
foyer
frail
frame
fray
fresh
fried
frill
frisk
from
front
frost
froth
frown
froze
fruit
gag
gains
gala
game
gap
gas
gave
gear
gecko
geek
gem
genre
gift
gig
gills
given
giver
glad
glass
glide
gloss
glove
glow
glue
goal
going
golf
gong
good
gooey
goofy
gore
gown
grab
grain
grant
grape
graph
grasp
grass
grave
gravy
gray
green
greet
grew
grid
grief
grill
grip
grit
groom
grope
growl
grub
grunt
guide
gulf
gulp
gummy
guru
gush
gut
guy
habit
half
halo
halt
happy
harm
hash
hasty
hatch
hate
haven
hazel
hazy
heap
heat
heave
hedge
hefty
help
herbs
hers
hub
hug
hula
hull
human
humid
hump
hung
hunk
hunt
hurry
hurt
hush
hut
ice
icing
icon
icy
igloo
image
ion
iron
islam
issue
item
ivory
ivy
jab
jam
jaws
jazz
jeep
jelly
jet
jiffy
job
jog
jolly
jolt
jot
joy
judge
juice
juicy
july
jumbo
jump
junky
juror
jury
keep
keg
kept
kick
kilt
king
kite
kitty
kiwi
knee
knelt
koala
kung
ladle
lady
lair
lake
lance
land
lapel
large
lash
lasso
last
latch
late
lazy
left
legal
lemon
lend
lens
lent
level
lever
lid
life
lift
lilac
lily
limb
limes
line
lint
lion
lip
list
lived
liver
lunar
lunch
lung
lurch
lure
lurk
lying
lyric
mace
maker
malt
mama
mango
manor
many
map
march
mardi
marry
mash
match
mate
math
moan
mocha
moist
mold
mom
moody
mop
morse
most
motor
motto
mount
mouse
mousy
mouth
move
movie
mower
mud
mug
mulch
mule
mull
mumbo
mummy
mural
muse
music
musky
mute
nacho
nag
nail
name
nanny
nap
navy
near
neat
neon
nerd
nest
net
next
niece
ninth
nutty
oak
oasis
oat
ocean
oil
old
olive
omen
onion
only
ooze
opal
open
opera
opt
otter
ouch
ounce
outer
oval
oven
owl
ozone
pace
pagan
pager
palm
panda
panic
pants
panty
paper
park
party
pasta
patch
path
patio
payer
pecan
penny
pep
perch
perky
perm
pest
petal
petri
petty
photo
plank
plant
plaza
plead
plot
plow
pluck
plug
plus
poach
pod
poem
poet
pogo
point
poise
poker
polar
polio
polka
polo
pond
pony
poppy
pork
poser
pouch
pound
pout
power
prank
press
print
prior
prism
prize
probe
prong
proof
props
prude
prune
pry
pug
pull
pulp
pulse
puma
punch
punk
pupil
puppy
purr
purse
push
putt
quack
quake
query
quiet
quill
quilt
quit
quota
quote
rabid
race
rack
radar
radio
raft
rage
raid
rail
rake
rally
ramp
ranch
range
rank
rant
rash
raven
reach
react
ream
rebel
recap
relax
relay
relic
remix
repay
repel
reply
rerun
reset
rhyme
rice
rich
ride
rigid
rigor
rinse
riot
ripen
rise
risk
ritzy
rival
river
roast
robe
robin
rock
rogue
roman
romp
rope
rover
royal
ruby
rug
ruin
rule
runny
rush
rust
rut
sadly
sage
said
saint
salad
salon
salsa
salt
same
sandy
santa
satin
sauna
saved
savor
sax
say
scale
scam
scan
scare
scarf
scary
scoff
scold
scoop
scoot
scope
score
scorn
scout
scowl
scrap
scrub
scuba
scuff
sect
sedan
self
send
sepia
serve
set
seven
shack
shade
shady
shaft
shaky
sham
shape
share
sharp
shed
sheep
sheet
shelf
shell
shine
shiny
ship
shirt
shock
shop
shore
shout
shove
shown
showy
shred
shrug
shun
shush
shut
shy
sift
silk
silly
silo
sip
siren
sixth
size
skate
skew
skid
skier
skies
skip
skirt
skit
sky
slab
slack
slain
slam
slang
slash
slate
slaw
sled
sleek
sleep
sleet
slept
slice
slick
slimy
sling
slip
slit
slob
slot
slug
slum
slurp
slush
small
smash
smell
smile
smirk
smog
snack
snap
snare
snarl
sneak
sneer
sniff
snore
snort
snout
snowy
snub
snuff
speak
speed
spend
spent
spew
spied
spill
spiny
spoil
spoke
spoof
spool
spoon
sport
spot
spout
spray
spree
spur
squad
squat
squid
stack
staff
stage
stain
stall
stamp
stand
stank
stark
start
stash
state
stays
steam
steep
stem
step
stew
stick
sting
stir
stock
stole
stomp
stony
stood
stool
stoop
stop
storm
stout
stove
straw
stray
strut
stuck
stud
stuff
stump
stung
stunt
suds
sugar
sulk
surf
sushi
swab
swan
swarm
sway
swear
sweat
sweep
swell
swept
swim
swing
swipe
swirl
swoop
swore
syrup
tacky
taco
tag
take
tall
talon
tamer
tank
taper
taps
tarot
tart
task
taste
tasty
taunt
thank
thaw
theft
theme
thigh
thing
think
thong
thorn
those
throb
thud
thumb
thump
thus
tiara
tidal
tidy
tiger
tile
tilt
tint
tiny
trace
track
trade
train
trait
trap
trash
tray
treat
tree
trek
trend
trial
tribe
trick
trio
trout
truce
truck
trump
trunk
try
tug
tulip
tummy
turf
tusk
tutor
tutu
tux
tweak
tweet
twice
twine
twins
twirl
twist
uncle
uncut
undo
unify
union
unit
untie
upon
upper
urban
used
user
usher
utter
value
vapor
vegan
venue
verse
vest
veto
vice
video
view
viral
virus
visa
visor
vixen
vocal
voice
void
volt
voter
vowel
wad
wafer
wager
wages
wagon
wake
walk
wand
wasp
watch
water
wavy
wheat
whiff
whole
whoop
wick
widen
widow
width
wife
wifi
wilt
wimp
wind
wing
wink
wipe
wired
wiry
wise
wish
wispy
wok
wolf
womb
wool
woozy
word
work
worry
wound
woven
wrath
wreck
wrist
xerox
yahoo
yam
yard
year
yeast
yelp
yield
yo-yo
yodel
yoga
yoyo
yummy
zebra
zero
zesty
zippy
zone
zoom


lib/CtrlO/Crypt/XkcdPassword.pm
package CtrlO::Crypt::XkcdPassword;
use strict;
use warnings;
use 5.010;

# ABSTRACT: Yet another xkcd style password generator

# VERSION

use Carp qw(croak);
use Crypt::Rijndael;
use Crypt::URandom;
use Data::Entropy qw(with_entropy_source);
use Data::Entropy::Algorithms qw(rand_int pick_r shuffle_r choose_r);
use Data::Entropy::RawSource::CryptCounter;
use Data::Entropy::Source;
use Data::Handle;
use Module::Runtime qw(use_module);

use base 'Class::Accessor::Fast';

__PACKAGE__->mk_accessors(qw(entropy wordlist language _list _pid));

=method new

  my $pw_generator = CtrlO::Crypt::XkcdPassword->new;

Initialize a new object. Uses C<CtrlO::Crypt::XkcdPassword::Wordlist::en_gb>
as a word list per default. The default entropy is based on
C<Crypt::URandom>, i.e. C</dev/urandom> and should be random enough (at
least more random than plain old C<rand()>).

If you want / need to supply another source of entropy, you can do so
by setting up an instance of C<Data::Entropy::Source> and passing it
to C<new> as C<entropy>.

  my $pw_generator = CtrlO::Crypt::XkcdPassword->new(
      entropy => Data::Entropy::Source->new( ... )
  );

To use one of the included language-specific word lists, do:

  my $pw_generator = CtrlO::Crypt::XkcdPassword->new(
      language => 'en-GB',
  );

Available languages are:

=over

=item * en-GB

=back

You can also provide your own custom word list, either in a file:

  my $pw_generator = CtrlO::Crypt::XkcdPassword->new(
      wordlist => '/path/to/file'
  );

Or in a module:

  my $pw_generator = CtrlO::Crypt::XkcdPassword->new(
      wordlist => 'My::Wordlist'
  );

See L</"DEFINING CUSTOM WORD LISTS"> for more info.

=cut

sub new {
    my ( $class, %args ) = @_;

    my %object;

    # init the word list
    my @list;
    if ( $args{wordlist} ) {
        $object{wordlist} = $args{wordlist};
    }
    else {
        my $lang = lc( $args{language} || 'en-GB' );
        $lang =~ s/-/_/g;
        $object{wordlist} = 'CtrlO::Crypt::XkcdPassword::Wordlist::' . $lang;
    }

    if ( -r $object{wordlist} ) {
        open( my $fh, '<:encoding(UTF-8)', $object{wordlist} );
        while ( my $word = <$fh> ) {
            chomp($word);
            $word =~ s/\s//g;
            push( @list, $word );
        }
        $object{_list} = \@list;
    }
    elsif ( $object{wordlist} =~ /::/ ) {
        _load_wordlist_from_package( \%object );
    }
    elsif ( $object{wordlist} =~ /^\w+$/i ) {
        $object{wordlist} = 'CtrlO::Crypt::XkcdPassword::Wordlist::' . $object{wordlist};
        _load_wordlist_from_package( \%object );
    }
     else {
        croak(    'Invalid word list: >'
                . $object{wordlist}
                . '<. Has to be either a Perl module or a file' );
    }

    # poor person's lazy_build
    $object{entropy} = $args{entropy} || $class->_build_entropy;
    $object{_pid} = $$;

    return bless \%object, $class;
}

sub _load_wordlist_from_package {
    my ( $object ) = @_;
    eval { use_module( $object->{wordlist} ); };
    if ($@) {
        croak( "Cannot load word list module " . $object->{wordlist} );
    }
    my $pkg = $object->{wordlist};
    no strict 'refs';

    # do we have a __DATA__ section, indication a subclass of https://metacpan.org/release/WordList
    my $handle = eval { Data::Handle->new($pkg) };
    if ($handle) {
        $object->{_list} = [ map { s/\n//g; chomp; $_ } $handle->getlines ];
    }
    # do we have @Words, indication Crypt::Diceware
    elsif ( @{"${pkg}::Words"} ) {
        $object->{_list} = \@{"${pkg}::Words"};
    }
    else {
        croak("Cannot find word list in Perl module $pkg");
    }
}

sub _build_entropy {
    my $class = shift;
    return Data::Entropy::Source->new(
        Data::Entropy::RawSource::CryptCounter->new(
            Crypt::Rijndael->new( Crypt::URandom::urandom(32) )
        ),
        "getc"
    );
}

=method xkcd

  my $pw = $pw_generator->xkcd;
  my $pw = $pw_generator->xkcd( words  => 3 );
  my $pw = $pw_generator->xkcd( digits => 2 );

Generate a random, xkcd-style password.

Per default will return 4 randomly chosen words from the word list,
each word's first letter turned to upper case, and concatenated
together into one string:

  $pw_generator->xkcd;
  # CorrectHorseBatteryStaple

You can get a different number of words by passing in C<words>. But
remember that anything smaller than 3 will probably make for rather
poor passwords, and anything bigger than 7 will be hard to remember.

You can also pass in C<digits> to append a random number consisting of
C<digits> digits to the password:

  $pw_generator->xkcd( words => 3, digits => 2 );
  # StapleBatteryCorrect75

=cut

sub xkcd {
    my ( $self, %args ) = @_;
    if ( $self->_pid != $$ ) {
        $self->_reinit_after_fork;
    }

    foreach my $key (keys %args) {
        croak "Invalid key [$key] received."
            unless ($key eq 'words' || $key eq 'digits');

        if (defined $args{$key} && ($args{$key} !~ /^[0-9]+$/)) {
            croak "Invalid value [$args{$key}] for key [$key].";
        }
    }

    my $word_count = $args{words} || 4;

    my $words = with_entropy_source(
        $self->entropy,
        sub {
            shuffle_r( choose_r( $word_count, $self->_list ) );
        }
    );

    if ( my $d = $args{digits} ) {
        push(
            @$words,
            sprintf(
                '%0' . $d . 'u',
                with_entropy_source(
                    $self->entropy, sub { rand_int( 10**$d ) }
                )
            )
        );
    }
    return join( '', map {ucfirst} @$words );
}

sub _reinit_after_fork {
    my $self = shift;
    $self->_pid($$);
    $self->entropy( $self->_build_entropy );
}

'correct horse battery staple';

__END__

=head1 SYNOPSIS

  use CtrlO::Crypt::XkcdPassword;
  my $password_generator = CtrlO::Crypt::XkcdPassword->new;

  say $password_generator->xkcd;
  # LimousineAllegeClergymanEconomic

  say $password_generator->xkcd( words => 3 );
  # ObservantFiresideMacho

  say $password_generator->xkcd( words => 3, digits => 3 );
  # PowerfulSpreadScarf645

  # Use custom word list
  CtrlO::Crypt::XkcdPassword->new(
    wordlist => '/path/to/wordlist'
  );
  CtrlO::Crypt::XkcdPassword->new(
    wordlist => 'Some::Wordlist::From::CPAN'
  );

  # Use another source of randomness (aka entropy)
  CtrlO::Crypt::XkcdPassword->new(
    entropy => Data::Entropy::Source->new( ... );
  );

=head1 DESCRIPTION

C<CtrlO::Crypt::XkcdPassword> generates a random password using the
algorithm suggested in L<https://xkcd.com/936/>: It selects 4 words
from a curated list of words and combines them into a hopefully easy
to remember password (actually a passphrase, but we're all trying to
get things done, so who cares..).

See L<this
explaination|https://www.explainxkcd.com/wiki/index.php/936:_Password_Strength>
for detailed information on the security of passwords generated from a
known word list.

But L<https://xkcd.com/927/> also applies to this module, as there are
already a lot of modules on CPAN implementing
L<https://xkcd.com/936/>. We still wrote a new one, mainly because we
wanted to use a strong source of entropy and a fine-tuned word list.

=head1 DEFINING CUSTOM WORD LISTS

Please note that C<language> is only supported for the default word list
included in this distribution.

=head2 in a plain file

Put your word list into a plain file, one line per word. Install this
file somewhere on your system. You can now use your word list like
this:

  CtrlO::Crypt::XkcdPassword->new(
    wordlist => '/path/to/wordlist'
  );

=head2 in a Perl module using the Wordlist API

L<Perlancar|https://metacpan.org/author/PERLANCAR> came up with a unified API for various word list modules,
implemented in L<Wordlist|https://metacpan.org/pod/WordList>. Pack
your list into a module adhering to this API, install the module, and
load your word list:

  CtrlO::Crypt::XkcdPassword->new(
    wordlist => 'Your::Cool::Wordlist'
  );

You can check out L<CtrlO::Crypt::XkcdPassword::Wordlist::en_gb> (included in
this distribution) for an example. But it's really quite simple: Just
subclass C<Wordlist> and put your list of words into the C<__DATA__>
section of the module, one line per word.

=head2 in a Perl module using the Crypt::Diceware API

David Golden uses a different API in his L<Crypt::Diceware> module,
which inspired the design of L<CtrlO::Crypt::XkcdPassword>. To use one
of those word lists, use:

  CtrlO::Crypt::XkcdPassword->new(
    wordlist => 'Crypt::Diceware::Wordlist::Common'
  );

(yes, this looks just like when using C<Wordlist>. We inspect the
wordlist module and try to figure out what kind of API you're using)

To create a module using the L<Crypt::Diceware> wordlist API, just
create a package containing a public array C<@Words> containing your
word list.

=head1 INCLUDED WORD LISTS

This distribution comes with a hand-crafted word list
L<CtrlO::Crypt::XkcdPassword::Wordlist::en_gb> and three word lists
provided by
L<EFF|https://www.eff.org/deeplinks/2016/07/new-wordlists-random-passphrases>:
L<CtrlO::Crypt::XkcdPassword::Wordlist::eff_large>,
L<CtrlO::Crypt::XkcdPassword::Wordlist::eff_short_1> and
L<CtrlO::Crypt::XkcdPassword::Wordlist::eff_short_2_0>.

=head1 WRAPPER SCRIPT

This distributions includes a simple wrapper script, L<pwgen-xkcd.pl>.

=head1 RUNNING FROM GIT

This is B<not> the recommended way to install / use this module. But
it's handy if you want to submit a patch or play around with the code
prior to a proper installation.

=head2 Carton

  git clone git@github.com:domm/CtrlO-Crypt-XkcdPassword.git
  carton install
  carton exec perl -Ilib -MCtrlO::Crypt::XkcdPassword -E 'say CtrlO::Crypt::XkcdPassword->new->xkcd'

=head2 cpanm & local::lib

  git clone git@github.com:domm/CtrlO-Crypt-XkcdPassword.git
  cpanm -L local --installdeps .
  perl -Mlocal::lib=local -Ilib -MCtrlO::Crypt::XkcdPassword -E 'say CtrlO::Crypt::XkcdPassword->new->xkcd'

=head1 SEE ALSO

Inspired by L<https://xkcd.com/936/> and L<https://xkcd.com/927/>.

There are a lot of similar modules on CPAN, so we just point you to
L<Neil Bower's comparison of CPAN modules for generating passwords|http://neilb.org/reviews/passwords.html>.

=head2 But why did we write yet another module?

=over

=item * Good entropy

Most of the password generating modules just use C<rand()>, which "is
not cryptographically secure" (according to perldoc).
C<CtrlO::Crypt::XkcdPassword> uses L<Crypt::URandom> via
L<Data::Entropy>, which provides good entropy while still being portable.

=item * Good word list

While L<Crypt::Diceware> has good entropy, we did not like its word
lists. Of course we could have just provided a word list better suited
to our needs, but we wanted it to be very easy to generate xkcd-style
passwords.

=item * Easy API

C<< my $pwd = CtrlO::Crypt::XkcdPassword->new->xkcd >> returns 4 words
starting with an uppercase letter as a string, which is our main use
case. Nevertheless, the API also allows for more or fewer words, or
even some digits.

=item * Fork save

=item * L<https://xkcd.com/927/>

=back

=head1 THANKS

=over

=item * Thanks to L<Ctrl O|http://www.ctrlo.com/> for funding the development of this module.

=item * We learned the usage of C<Data::Entropy> from
L<https://metacpan.org/pod/Crypt::Diceware>, which also implements an
algorithm to generate a random passphrase.

=item * L<m_ueberall|https://twitter.com/m_ueberall/status/965263922310909952>
for pointing out
L<https://www.explainxkcd.com/wiki/index.php/936:_Password_Strength>

=back
